### PR TITLE
Make AppClass::transport_generator initialized at const-init time

### PR DESCRIPTION
**Pull request by RedBeard0531**
## What, How & Why?
<!-- Describe the changes and give some hints to guide your reviewers if possible. -->
<!-- E.g. reference to other repos: This closes realm/realm-sync#??? -->

This fixes the "bad_function_call" issue where it was invoked while null                                                                                                                                                                                             (probably actually uninitialized). I'm not sure why it wasn't                                                                                                                                                                                                        initialized yet, since the compiler should have ensured that it was.                                                                                                                                                                                                 There is likely a compiler bug there but I'll need to dig in more. To                                                                                                                                                                                                work around that, I made it a simple function pointer, which lambdas can                                                                                                                                                                                             implicitly convert to, and all of that should be possible at const-init                                                                                                                                                                                              time (basically, it should be burned into the binary image with nothing                                                                                                                                                                                              happening at runtime). Unfortunately due to a different compiler bug,                                                                                                                                                                                                (https://developercommunity.visualstudio.com/t/const-init-of-function-pointers-from-lambdas/1383098),                                                                                                                                                                this requires putting a `+` in front of the lambda to explicitly decay                                                                                                                                                                                               it to a function pointer, rather than relying on the implicit conversion.  


## ☑️ ToDos
<!-- Add your own todos here -->
* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 🚦 Tests
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

**Review by kneth — COMMENTED** [review]
I think it is a bit over my C++ skills. Feel free to merge when CI passes.

File: src/js_network_transport.hpp [hunk 1]
```diff
@@ -179,7 +179,7 @@ struct JavaScriptNetworkTransport : public app::GenericNetworkTransport {
     }
 
     using SendRequestHandler = void(ContextType m_ctx, const app::Request request, std::function<void(const app::Response)> completion_callback);
-    using NetworkTransportFactory = std::function<std::unique_ptr<app::GenericNetworkTransport>(ContextType, Dispatcher)>;
+    using NetworkTransportFactory = std::unique_ptr<app::GenericNetworkTransport>(*)(ContextType, Dispatcher);
```
Comment (fronck) [inline src/js_network_transport.hpp:182]: We should probably leave a comment here as to why the C-style function pointer is necessary.  Just so someone doesn't "clean up" the code and re-introduce the bug.

**Review by fronck — CHANGES_REQUESTED** [review]
Please leave a couple of comments for posterity about why the workaround is here, so we don't inadvertently reverse it.

Otherwise, LGTM.

File: src/js_network_transport.hpp (hunk 1 above)
Comment (RedBeard0531) [reply src/js_network_transport.hpp:184]: :+1:

**Pull request merged**